app/adapters/signal_adapter/inbound.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+module SignalAdapter
+  UNKNOWN_CONTRIBUTOR = :unknown_contributor
+  UNKNOWN_CONTENT = :unknown_content
+
+  class Inbound
+    UNKNOWN_CONTENT_KEYS = %w[mentions attachments contacts].freeze
+
+    attr_reader :sender, :text, :message
+
+    def initialize
+      @callbacks = {}
+    end
+
+    def on(callback, &block)
+      @callbacks[callback] = block
+    end
+
+    def consume(signal_message)
+      signal_message = signal_message.with_indifferent_access
+
+      @text = signal_message.dig(:envelope, :dataMessage, :message)
+      @sender = initialize_sender(signal_message)
+      return unless @sender
+
+      @message = initialize_message(signal_message)
+      return unless @message
+
+      yield(@message) if block_given?
+    end
+
+    private
+
+    def trigger(event, *args)
+      return unless @callbacks.key?(event)
+
+      @callbacks[event].call(*args)
+    end
+
+    def initialize_sender(signal_message)
+      phone_number = signal_message.dig(:envelope, :source)
+      sender = Contributor.find_by(phone_number: phone_number)
+      return sender if sender
+
+      trigger(UNKNOWN_CONTRIBUTOR, phone_number) and return nil
+    end
+
+    def initialize_message(signal_message)
+      is_receipt = signal_message.dig(:envelope, :receiptMessage)
+      return nil if is_receipt
```

Comment:
We should probably handle other message types like `reaction`s or perhaps `remoteDelete` messages equally

Reply:
Ignore `remoteDelete` :smiling_imp: :wink: